File: android/edge-frameworks-core/src/main/kotlin/io/github/rajibsarwar/edgeframeworks/EdgeModelProvider.kt
package io.github.rajibsarwar.edgeframeworks

import kotlinx.coroutines.flow.Flow

interface EdgeModelProvider {
    val id: String

    suspend fun capabilities(): Set<EdgeCapability>

    suspend fun generate(
        request: EdgeGenerationRequest
    ): EdgeGenerationResponse

    fun stream(
        request: EdgeGenerationRequest
    ): Flow<EdgeGenerationEvent>
}

File: ios/Sources/EdgeFrameworks/Core/EdgeModelProvider.swift
import Foundation

public protocol EdgeModelProvider: Sendable {
    var id: String { get }

    func capabilities() async -> Set<EdgeCapability>

    func generate(_ request: EdgeGenerationRequest) async throws -> EdgeGenerationResponse

    func stream(_ request: EdgeGenerationRequest) -> AsyncThrowingStream<EdgeGenerationEvent, Error>
}
